akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala

```diff
@@ -290,13 +289,18 @@ private[http] object OutgoingConnectionBlueprint {
       // which builds a cache of all header instances seen on that connection
       val parser = rootParser.createShallowCopy()
       var waitingForMethod = true
+      val responseContexts = new java.util.ArrayDeque[BypassData]()
 
       setHandler(bypassInput, new InHandler {
         override def onPush(): Unit = {
           val responseContext = grab(bypassInput)
-          parser.setContextForNextResponse(responseContext)
-          val output = parser.parseBytes(ByteString.empty)
-          drainParser(output)
+          responseContexts.add(responseContext)
+          if (responseContexts.size() == 1) {
+            parser.setContextForNextResponse(responseContext)
+            val output = parser.parseBytes(ByteString.empty)
+            drainParser(output)
+          }
+          pull(bypassInput)
```

Comment:
Could we just limit the size here to the pipelineLimit?

    if (responseContexts.size() < pipelineLimit) pull(bypassInput)